# Introduction to Python Jupyter Notebooks
<font size="1" color=lightgray>Portions of the basic Python instructions in this notebook are taken from the open-source Python In Chemistry project (<a href="http://pythoninchemistry.org/" style="color:lightgray">http://pythoninchemistry.org</a>) by Fiona Dickinson, Ben Morgan & Andrew McCluskey.</font>

(This notebook makes some use of LaTeX markup; you may wish to complete the LaTeX part of HW1 first if their use is unclear.)

Python is currently one of the most commonly used programming languages for analysis in any technical field. It is extremely versatile, and can be used for nearly anything. I personally use it in my astrophysics research all the time, as well as for all sorts of other things (for example, once, when looking for a place to rent in a different city, I used it to fetch all of the craigslist ads, filter out ones that met my criteria, and return an organized, filtered list of what was available so that I didn't have to sort through hundreds of listings). In PH304, you learned how to use Python to solve some physics problems. In that class you used a particular implementation of Python called VPython, so-called because it helped you visualize things.

In this class we will be using a different implementation called Jupyter notebooks (which is what you're looking at right now!). These let you use Python in a notebook-like way - you can write prose (like I'm doing now) and mix it in with working code. This isn't so good for giant simulations or corporate code worked on by hundreds of people (you'd want a less interactive implementation for that!), but it is great for taking notes, doing little test calculations, and showing other people how to do things. **Please work through this notebook, completing the parts that ask you to do so.**

## Cells
#### Markdown Cells
The first thing to notice is that this notebook is comprised of "cells." When you click on a chunk of text or code, you select the cell. A button pops up that you can click to edit the contents of the cell. This cell is called a "markdown" cell, because it uses the "markdown" language to format prose. It is similar to how text is formated on many websites. You can make things **bold** or *italic*. You can use LaTeX-style equation writing: $E=mc^2$. When you are done editing a cell, you click the "Run" button above to execute the markdown that you wrote, turning it into pretty text. **Please edit this cell and replace the following word with the (correctly typeset) equation for the area of a circle:** defenestrate. Run the cell to check your handiwork.

You can add a new cell with the button that looks like a plus sign at the top left of your browser. The button adds a cell below the currently selected cell. By default, it will be a code cell, but you can use the pull-down menu to change it to markdown. **Please add a new markdown cell below this one, and use it to describe how much coding experience you have (if any) and what programming languages you have experiences with.** Your cell should have the header "My Programming Experience" - this should be correctly formatted as a header by prefacing it with the # sign, as I did in this cell. Run the cell, checking that everything is formatted correctly.

#### Code Cells

When you want to actually do a calculation instead of write prose, add a code cell. Below, I have added a very simple code cell that adds two and five and multiplies the result by ten:

In [1]:
(2+5)*10

70

When the cell is run, it displays the output "70" as expected. **Below this cell, please add a code cell that subtracts 4.7 from 10 and multiplies the result by 2, and verify that it outputs the expected result.**

## Functions and Variables

More commonly, we want to do more complicated calculations, which usually involves defining variables and functions. A variable is kinda like an algebraic variable - it's a thing that refers to some actual value. For example, I can create a variable called <code>a</code> and assign it the value 3. Then, if I tell the notebook to print the variable <code>a</code>, it will output the numerical value 3. I do this below:

In [2]:
a=3
print(a)

3


A function is just a little program that operates on variables and returns a result. For example, I can write a function that evaluates $x^2$:

In [3]:
def f(x):
    y = x ** 2
    return y

What this says is that you want to: 
- <code>def</code>ine a function
- That function should be called <code>f</code>
- It will have one argument, <code>x</code>
- The operation to be performed on <code>x</code> is to square <code>x</code>; <code>x ** 2</code>
- The result of the operation should be stored in the variable <code>y</code>
- The data stored in variable <code>y</code> should be <code>return</code>ed

Be sure to notice that in Python, whitespace is important. The code that is inside the function must be consistently indented.

Defining a function is only half the battle, now we need to *call* it. This is achieved using the following *syntax*:

<code>result = name_of_function(argument)
</code>

This sort of code tells the computer to *execute* the function called <code>name_of_function</code>, with the argument <code>argument</code>, and store the result in the variable <code>result</code>. 

So if we would like to run the function that we have defined above, we could do the following, in which I have used ridiculous variable names to illustrate that they can be anything:

In [4]:
pig = 3
shark = f(pig)
shark

9

Keep in mind that I must run the code cell that defines the function prior to using it in subsequent code cells.

I can also define a function and use it all in one code cell, have more than one argument, and operate on things other than numbers, like so:

In [5]:
def mySuperDuperFunction(boringWord,numSupers):
    excitingWord = (numSupers*'super-')+boringWord
    return excitingWord

fixedWord = mySuperDuperFunction('aardvarks',3)
print(fixedWord)

super-super-super-aardvarks


**Please add a code cell below that defines a function that evaluates $x^n$ (i.e., your function should take two arguments and be capable of evaluating any combination of $x$ and $n$). Use that function to calculate the value of $(2.7)^{3.4}$, store it in a variable called `gorgonzola`, and print that variable once it contains the correct result.**

## Working with physical constants and units, with `scipy.constants`

In this calss, most of the calculations of interest involve physical constants, such as the speed of light, $c$, Avogadro's constant, $N_\mathrm{A}$, Planck's constant, $h$, and the Boltzmann constant, $k_\mathrm{B}$. When working with algebraic expressions, it is also important to keep track of the relevant units.

While manipulating units manually (for example, when working through a problem by hand) is an important skill, if you are solving numerical problems using code you ideally want to avoid typing in unit conversions and values for constants. Any numbers that are typed in, are possible places where you could mistype. Mistakes in your code can hopefully be spotted when they produce errors, or from testing that your code produces a known result. Mistakes in numbers typed in are harder to spot, and can propogate through to your results.

To reduce the chance of introducing errors, there are two easy techniques you can follow:

1. Define your constants and unit conversions once by assigning them to variables (conventionally at the top of your code), rather than retyping them each time.
2. If possible, read constants and unit conversions automatically from databases.

## An Example - calculating the area of a circle

**Q**. The equation for the area of a circle is:
\begin{equation}
A=\pi r^2.
\end{equation}
Calculate the area of a circle with radius 5.0 cm. Give your results to 3 decimal places.

##### Approach 1 — numbers are typed in-place (the worst approach - don't ever do this!)

In [6]:
area = 3.1415 * 5.0**2

print( "area = {:.3f} cm^2".format( area ) )

area = 78.538 cm^2


As an aside, the `.format` expression is a "method." It's one way of controlling the formatting of the output. It replaces the stuff in curly brackets with the listed variables, and you can control their formatting with codes inside the curly brackets. It's probably easier to look at an example than to read a description:

In [7]:
numberone=1
numbertwo=2
print('{} and {}'.format(numberone, numbertwo)) # print the two variables without formatting restrictions (python decides) (notice how I can add comments to my code)
print('{:06.3f} and {:06.3f}'.format(numberone, numbertwo)) # print the two variables using 6 character spaces, keeping leading zeroes, and with 3 decimal places (i.e., print them as floating point numbers)

1 and 2
01.000 and 02.000


##### Approach 2 — using variables (better, but not the best)

In [8]:
r = 5.0 # radius in cm.
pi = 3.1415 # pi to 4 decimal places

area = pi * r**2

print( "area = {:.3f} cm^2".format( area ) )

area = 78.538 cm^2


This approach has two advantages over the first. The variables `r` and `pi` are clearly defined at the top of the code, and are then referred to by their variable names in the actual calculations. If we go on to use these variables in a subsequent calculation, we can be sure that the same values of `r` and `pi` are used in both calculations, and it is a lot clearer where these numbers come from. By dividing up the code like this, it is also easier to understand. The actual calculations now look like the expression in the question, and it is clearer to someone reading the code what is happening.

##### Approach 3 — importing constants (ideal)

Instead of typing in the value of $\pi$, you can `import` it from the `math` module. One of Python's great strength is the ability to `import` code from modules, which can be tools that are built in to Python or tools that other people made. The `math` module contains basic constants and math functions, which are listed in its documentation: https://docs.python.org/3/library/math.html.

In [9]:
from math import pi
print( 'pi = {}'.format(pi) )

pi = 3.141592653589793


This means you never have to worry about mistyping the numerical value. You also get a more accurate value. Notice that in the previous examples, the last decimal place was incorrectly rounded to give `3.1415` and not `3.1416`, and this rounding error propagates through to the calculated answers.

In [10]:
from math import pi

# question inputs
r = 5.0 # radius in cm

area = pi * r**2

print( "area = {:.3f} cm^2".format( area ) )

area = 78.540 cm^2


## Working with `scipy.constants`

The `math` module contains various mathematical constants, such as $\pi$ and $\mathrm{e}$ (the natural logarithm base). For **physical constants** you can use the `scipy.constants` module: https://docs.scipy.org/doc/scipy/reference/constants.html

Common physical constants can be imported directly, e.g.

In [11]:
from scipy.constants import c # speed of light
from scipy.constants import h # Planck's constant
from scipy.constants import k # Boltzmann constant

print( 'c = {} m s^-1'.format( c ) )
print( 'h = {} J s'.format( h ) )
print( 'k = {} J K^-1'.format( k ) )

c = 299792458.0 m s^-1
h = 6.62607004e-34 J s
k = 1.38064852e-23 J K^-1


`scipy.constants` also contains a `physical_constants` database (as a Python dictionary) that contains an even larger number of constants, and also gives their units.

In [12]:
from scipy.constants import physical_constants

print( physical_constants['speed of light in vacuum'] )
print( physical_constants['Planck constant'] )
print( physical_constants['Boltzmann constant'] )

(299792458.0, 'm s^-1', 0.0)
(6.62607004e-34, 'J s', 8.1e-42)
(1.38064852e-23, 'J K^-1', 7.9e-30)


Each entry in the dictionary returns three components: the value, the units (as a string), and the experimental uncertainty.

## In practice — constants and unit conversions in homework

The following are example questions that involve working with different units, that illustrate using the approaches described above. I'll complete the first two questions in a way that would receive full credit on a homework assignment. **You must complete the third and fourth questions (Hint: format your code sorta similarly to my answer to the second question). You must do all numerical calculations within the code.** Your code should always be clearly commented.

### 1.

**Q** The pressure at the top of a mountain is one-third of an atmosphere. Calculate the pressure in S.I. units and in bar.

In [13]:
from scipy.constants import atm, bar # 1 atm and 1 bar in SI units (Pascals).

print( "standard atmosphere in Pa: {}".format( atm ) )
print( "one bar in Pa: {}".format( bar ) )

standard atmosphere in Pa: 101325.0
one bar in Pa: 100000.0


In [14]:
pressure_in_atm = 1/3
pressure_in_Pa = pressure_in_atm * atm
pressure_in_bar = pressure_in_Pa / bar

print( "pressure in S.I. units (Pa) = {}".format( pressure_in_Pa ) )
print( "pressure in bar = {}".format( pressure_in_bar ) )

pressure in S.I. units (Pa) = 33775.0
pressure in bar = 0.33775


### 2.

**Q** What is the volume of one mole of air, at room temperature and 1  atm pressure (problem 1.9 from Schroeder)? Solve this using Python code.

**A**. We can answer this question by applying the ideal gas law:

\begin{equation}
pV = nRT
\end{equation}
which can be rearranged to give
\begin{equation}
V = \frac{nRT}{P}
\end{equation}
We simply want to write code to evaluate this equation:

In [15]:
from scipy.constants import atm, R # 1 atm in SI units, molar gas constant in SI units

# Note that I didn't actually have to re-import R because I did it earlier in the notebook. I just did it here for clarity.

moles = 1.0 # given information, number of moles of gas
temperature = 300.0 # assume a room temperature in K

def idealGasVolume(n, T, P):
    return n*R*T/P

print( 'volume =', idealGasVolume(moles, temperature, atm), 'm^3' )

volume = 0.0246172014803849 m^3


### 3.

**Q** What is the temperature of one mole of an ideal gas, at 0.1 atm pressure, confined to 1 $\rm m^3$? **Solve this using Python code by adding one or more cells below this one.** Be sure to use sufficient markdown cells to explain your answer, as I did in the question above. (Hint: Useful scipy constants might be `nano` and `k`.)

### 4.

**Q**  What is the typical distance between neightboring molecules of air at STP (273 K and 1 atm)? How does this compare to the size of a molecule? Does this support or refute the claim that the ideal gas law is a reasonable approximation at most common Earth temperatures? (Hint: Once you have the average volume per molecule, what mathematical operation will change a volume scale to a length scale?) **Solve this using Python code by adding one or more cells below.** Be sure to use sufficient markdown cells to explain your answer, as I did above.